### PR TITLE
[REVIEW] fix `device_scalar` and its tests so that they use the correct CUDA stream

### PR DESCRIPTION
**Pull request by rongou**
While debugging #563, found a few issues with the `device_scalar` and its tests. But I don't think these fixes would help with the `cudaErrorIllegalAddress` problems in Spark.

The main thing is we need to use the correct CUDA stream for setting/getting the value, and copying the `device_scalar`. By default these methods use the null stream, but in the tests we create a new non-default stream when constructing the `device_scalar`. This is not a problem when using the legacy default stream which does implicit synchronization, but when running with per-thread default stream, the mismatched streams create race conditions that cause the tests to fail.

@jrhemstad @harrism

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: include/rmm/device_scalar.hpp [hunk 1]
```diff
@@ -82,6 +82,25 @@ class device_scalar {
     set_value(initial_value, stream);
   }
 
+  /**
+   * @brief Construct a new `device_scalar` by deep copying the contents of
+   * another `device_scalar`, optionally using the specified stream and memory
```
Comment (harrism) [inline include/rmm/device_scalar.hpp:87]: The stream and memory resource are always used, they just have defaults.

```suggestion
   * another `device_scalar`, using the specified stream and memory
```

File: include/rmm/device_scalar.hpp [hunk 2]
```diff
@@ -212,10 +231,9 @@ class device_scalar {
    */
   T const *data() const noexcept { return static_cast<T const *>(buffer.data()); }
 
-  device_scalar()                      = default;
-  ~device_scalar()                     = default;
-  device_scalar(device_scalar const &) = default;
```
Comment (harrism) [inline include/rmm/device_scalar.hpp:217]: What happened to this one?

**Review by harrism — CHANGES_REQUESTED** [review]
Can you please explain the problems that are being fixed in the issue description (and provide a little more detail in the title and changelog)?

File: include/rmm/device_scalar.hpp (hunk 2 above)
Comment (rongou) [reply include/rmm/device_scalar.hpp:217]: This was moved up to include the optional stream and mr parameters.

**Comment by rongou** [comment]
@harrism updated the PR title and changlog entry, also added some explanation to the problem in the description. Please take another look.

**Comment by harrism** [comment]
Ok, this points to the fact that we need to generate PTDS versions of all tests. Currently RMM is tested only with PTDS enabled. There is an open issue for this.

**Comment by rongou** [comment]
Do you mean having both the legacy and PTDS versions of tests? The test failures being fixed here only occur with PTDS, but not all the time. Need to do something like this to reliably reproduce:
```bash
for i in {1..100}; do ./gtests/DEVICE_SCALAR_TEST || break; done
```

**Comment by harrism** [comment]
Yeah like we do for POOL_MR tests and DEVICE_MR tests. See the test cmakelists.txt. I think we should basically make every test compiled twice -- once for legacyDS and once for PTDS. 

The problem is currently CI runs with PTDS on always (mistake, see #555 )

**Comment by rongou** [comment]
#603

**Comment by harrism** [comment]
Thanks!

**Review by harrism — APPROVED** [review]

**Comment by rongou** [comment]
@kkraus14 can this be merged? Thanks!

File: include/rmm/device_scalar.hpp [hunk 3]
```diff
@@ -82,6 +82,25 @@ class device_scalar {
...
+   * @throws rmm::bad_alloc If creating the new allocation fails.
+   * @throws rmm::cuda_error if copying from `other` fails.
+   *
+   * @param other The `device_scalar` whose contents will be copied
+   * @param stream The stream to use for the allocation and copy
+   * @param mr The resource to use for allocating the new `device_scalar`
+   */
+  device_scalar(device_scalar const &other,
+                cudaStream_t stream                 = 0,
+                rmm::mr::device_memory_resource *mr = rmm::mr::get_current_device_resource())
```
Comment (jrhemstad) [inline include/rmm/device_scalar.hpp:99]: This changes behavior. The original copy ctor would use the resource from `other`.

**Review by jrhemstad — CHANGES_REQUESTED** [review]

File: include/rmm/device_scalar.hpp (hunk 3 above)
Comment (rongou) [reply include/rmm/device_scalar.hpp:99]: Not exactly. The original copy ctor calls `device_buffer`'s copy ctor, which defaults to the null stream: https://github.com/rapidsai/rmm/blob/branch-0.17/include/rmm/device_buffer.hpp#L144

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**